File: solutions/Medium/Letter_Combinations_of_a_Phone_Number_0017.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        """
        Function: letterCombinations
        Description: Returns all possible letter combinations represented by a string of digits (2-9).

        Parameters:
        - digits (str): String containing digits (2-9).

        Returns:
        - List[str]: All possible letter combinations in any order.
        """

        if not digits:
            return []

        # Mapping of digits to corresponding letters
        digit_to_letters = {
            "2": "abc", "3": "def", "4": "ghi", "5": "jkl",
            "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"
        }

        res = []

        # Backtracking function
        def backtrack(index, path):
            if index == len(digits):
                res.append("".join(path))
                return

            for char in digit_to_letters[digits[index]]:
                backtrack(index + 1, path + [char])

        # Start backtracking from the first digit
        backtrack(0, [])

        return res


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: Standard case with multiple digits
    print("Test Case 1: Digits '23'")
    digits1 = "23"
    print("Expected Output: ['ad','ae','af','bd','be','bf','cd','ce','cf'] | Actual Output:",
          solution.letterCombinations(digits1))

    # Test case 2: Empty string input
    print("Test Case 2: Empty string")
    digits2 = ""
    print("Expected Output: [] | Actual Output:",
          solution.letterCombinations(digits2))

    # Test case 3: Single digit case
    print("Test Case 3: Single digit '2'")
    digits3 = "2"
    print("Expected Output: ['a', 'b', 'c'] | Actual Output:",
          solution.letterCombinations(digits3))

    # Test case 4: Largest input case with 4 digits
    print("Test Case 4: Maximum length input '7896'")
    digits4 = "7896"
    print("Expected Output: Computed Output:",
          solution.letterCombinations(digits4))

    # Test case 5: Digits containing 7 and 9 (which have 4 letters)
    print("Test Case 5: Digits '79'")
    digits5 = "79"
    print("Expected Output: Computed Output:",
          solution.letterCombinations(digits5))

    # Test case 6: Consecutive numbers
    print("Test Case 6: Digits '456'")
    digits6 = "456"
    print("Expected Output: Computed Output:",
          solution.letterCombinations(digits6))

"""
Time Complexity:
- O(4^N) → Each digit (2-9) maps to 3-4 characters, resulting in exponential complexity.
- Space Complexity:
  - O(N) → Recursion stack depth for backtracking.
"""

File: solutions/Medium/Letter_Combinations_of_a_Phone_Number_0017.js
/**
 * Function: letterCombinations
 * Description: Given a string of digits from 2-9, this function returns all possible letter 
 * combinations that the number could represent based on the telephone keypad mapping.
 *
 * @param {string} digits - A string containing digits from '2' to '9'.
 * @returns {string[]} - An array of all possible letter combinations.
 */
const letterCombinations = function(digits) {
    if (!digits.length) return []; // Return an empty array if input is empty.

    // Mapping of digits to corresponding letters on a telephone keypad.
    const digitToLetters = {
        '2': 'abc',
        '3': 'def',
        '4': 'ghi',
        '5': 'jkl',
        '6': 'mno',
        '7': 'pqrs',
        '8': 'tuv',
        '9': 'wxyz'
    };

    let result = [];

    /**
     * Backtracking helper function to generate all letter combinations.
     * @param {string} combination - Current string combination being built.
     * @param {number} index - Current index in the digits string.
     */
    const backtrack = (combination, index) => {
        // If combination length matches digits length, add to result.
        if (index === digits.length) {
            result.push(combination);
            return;
        }

        // Retrieve the corresponding letters for the current digit.
        let letters = digitToLetters[digits[index]];
        for (let letter of letters) {
            backtrack(combination + letter, index + 1);
        }
    };

    // Start the backtracking process.
    backtrack("", 0);

    return result;
};

// Example Test Cases
if (require.main === module) {
    console.log("Test 1: Basic case with two digits");
    console.log(letterCombinations("23"));
    // Expected output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"]

    console.log("Test 2: Edge case - empty input");
    console.log(letterCombinations(""));
    // Expected output: []

    console.log("Test 3: Single digit input");
    console.log(letterCombinations("2"));
    // Expected output: ["a", "b", "c"]

    console.log("Test 4: Two-digit input with four-letter mappings");
    console.log(letterCombinations("79"));
    // Expected output: ["pw", "px", "py", "pz", "qw", "qx", "qy", "qz", "rw", "rx", "ry", "rz", "sw", "sx", "sy", "sz"]

    console.log("Test 5: Three-digit input");
    console.log(letterCombinations("345"));
    // Expected output: 27 combinations of 'd', 'e', 'f' with 'g', 'h', 'i' with 'j', 'k', 'l'

    console.log("Test 6: Four-digit input (maximum allowed)");
    console.log(letterCombinations("6789"));
    // Expected output: 3x3x4x4 = 144 combinations
}

// Complexity Analysis
// Time Complexity: O(4^n) - Each digit has 3 or 4 letters leading to exponential growth.
// Space Complexity: O(4^n) - Stores all combinations in the result array.

export default letterCombinations;
